common/authentication/aws/x509_iam.go

```
@@ -0,0 +1,430 @@
+	var x509Auth x509TempAuth
+	if err := kitmd.DecodeMetadata(opts.Properties, &x509Auth); err != nil {
+		return err
+	}
+
+	switch {
+	case x509Auth.TrustProfileArn == nil:
+		return errors.New("trustProfileArn is required")
+	case x509Auth.TrustAnchorArn == nil:
+		return errors.New("trustAnchorArn is required")
+	case x509Auth.AssumeRoleArn == nil:
+		return errors.New("assumeRoleArn is required")
+	case x509Auth.SessionDuration == nil:
+		awsDefaultDuration := time.Duration(900) // default 15m
+		x509Auth.SessionDuration = &awsDefaultDuration
+	}
+
+	a.logger = opts.Logger
+	a.TrustProfileArn = x509Auth.TrustProfileArn
+	a.TrustAnchorArn = x509Auth.TrustAnchorArn
+	a.AssumeRoleArn = x509Auth.AssumeRoleArn
+	a.SessionDuration = x509Auth.SessionDuration
+	a.cfg = GetConfig(opts)
+	a.clients = newClients()
+
+	err := a.getCertPEM(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get x.509 credentials: %v", err)
+	}
+
+	// Parse trust anchor and profile ARNs
+	if err := a.initializeTrustAnchors(); err != nil {
+		return err
+	}
+
+	initialSession, err := a.createOrRefreshSession(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to create the initial session: %v", err)
+	}
+	a.session = initialSession
+	go a.startSessionRefresher(context.Background())
+
+	return nil
+}
+
+func (a *x509TempAuth) Close() error {
+	return nil
+}
+
+func (a *x509TempAuth) getCertPEM(ctx context.Context) error {
+	// retrieve svid from spiffe context
+	svid, ok := spiffecontext.From(ctx)
+	if !ok {
+		return fmt.Errorf("no SVID found in context")
+	}
+	// get x.509 svid
+	svidx, err := svid.GetX509SVID()
+	if err != nil {
+		return err
+	}
+
+	// marshal x.509 svid to pem format
+	chainPEM, keyPEM, err := svidx.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed to marshal SVID: %w", err)
+	}
+
+	a.chainPEM = chainPEM
+	a.keyPEM = keyPEM
+	return nil
+}
+
+func (a *x509TempAuth) S3(ctx context.Context) *S3Clients {
```

All of these funcs should return an error which happens when the given context is cancelled. We wait for ctx cancelled or x509TempAuth is ready.

_or_ we don't need to give a ctx at all if a.clients is never nil.